---
title: "Single Cell RNAseq Part 3"
author: "Bioinformatics Core"
output:
    html_document:
      keep_md: TRUE
---
## Load libraries
```{r, warning=FALSE,error=FALSE,message=FALSE}
library(Seurat)
```

## Load the Seurat object
```{r}
load(file="pre_sample_corrected.RData")
experiment.aggregate
experiment.test <- experiment.aggregate
set.seed(12345)
rand.genes <- sample(1:nrow(experiment.test), 500,replace = F)
mat <- as.matrix(GetAssayData(experiment.test, slot="data"))
mat[rand.genes,experiment.test$batchid=="Batch2"] <- mat[rand.genes,experiment.test$batchid=="Batch2"] + 0.22
experiment.test = SetAssayData(experiment.test, slot="data", new.data= mat )
```

## Exploring Batch effects, none, Seurat [vars.to.regress]

First lets view the data without any corrections

## PCA in prep for tSNE

ScaleData - Scales and centers genes in the dataset.
```{r}
?ScaleData
experiment.test.noc <- ScaleData(object = experiment.test)
```

### Run PCA
```{r, fig.align="center", fig.caption="Uncorrected PCA plot"}
experiment.test.noc <- RunPCA(object = experiment.test.noc)
DimPlot(object = experiment.test.noc, group.by = "batchid", reduction = "pca")
DimPlot(object = experiment.test.noc, group.by = "batchid", dims = c(2,3), reduction = "pca")
```

PCA Elbow plot to determine how many principal components to use in downstream analyses.  Components after the "elbow" in the plot generally explain little additional variability in the data.

```{r}
ElbowPlot(experiment.test.noc)
```

We use 10 components in downstream analyses. Using more components more closely approximates the full data set but increases run time.

### TSNE Plot
```{r, fig.align="center", fig.caption="Uncorrected TSNE plot"}
pcs.use <- 10
experiment.test.noc <- RunTSNE(object = experiment.test.noc, dims = 1:pcs.use)
DimPlot(object = experiment.test.noc,  group.by = "batchid")

```

## Correct for sample to sample differences (seurat)

Use vars.to.regress to correct for the sample to sample differences and percent mitochondria
```{r, fig.align="center", fig.caption="Corrected PCA Plot"}
experiment.test.regress <- ScaleData(object = experiment.test,
                    vars.to.regress = c("batchid"), model.use = "linear")

experiment.test.regress <- RunPCA(object =experiment.test.regress)

DimPlot(object = experiment.test.regress, group.by = "batchid", reduction = "pca")
```

### Corrected TSNE Plot
```{r, fig.align="center", fig.caption="Corrected TSNE Plot"}
experiment.test.regress <- RunTSNE(object = experiment.test.regress, dims.use = 1:pcs.use)
DimPlot(object = experiment.test.regress, group.by = "batchid", reduction = "tsne")
```

#### Question(s)

1. Try a couple of PCA cutoffs (low and high) and compare the TSNE plots from the different methods.  Do they look meaningfully different?

## Get the next Rmd file
```{r, eval=FALSE}
download.file("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2020-August-intro-scRNAseq/master/data_analysis/scRNA_Workshop-PART4.Rmd", "scRNA_Workshop-PART4.Rmd")
```

## Session Information
```{r}
sessionInfo()
```
